File: data-stuctures/Stack.py
class Node:
    def __init__(self, val: int, prev=None):
        self.val = val
        self.prev = prev

class Stack:
    def __init__(self):
        self.head = None
        self.length = 0

    def push(self, val: int):
        node = Node(val= val)
        self.length += 1
        if not self.head:
            self.head = node
        else:
            node.prev = self.head
            self.head = node

    def pop(self):
        if self.head:
            self.length -= 1
            self.head = self.head.prev  

    def top(self):
        if self.head:
            return self.head.val
        return None
        


# Your MinStack object will be instantiated and called as such:
obj = Stack()
obj.push(5)
obj.push(7)
obj.push(6)
obj.push(4)
obj.pop()
obj.pop()
param_3 = obj.top()
print(obj.length)

File: data-stuctures/dart/Stack.dart
class Node<T> {
  T value;
  Node<T>? prev;

  Node(this.value);
}

class Stack<T> {
  int _length = 0;

  Node<T>? _head;

  Stack() {
    this._head = null;
  }

  int get length => _length;

  void push(T item) {
    _length++;
    Node<T> node = Node(item);
    if (_head == null) {
      _head = node;
    } else {
      node.prev = _head;
      _head = node;
    }
  }

  T? pop() {
    if (_head == null) {
      return null;
    } else {
      _length--;
      Node<T> head = _head!;
      _head = _head?.prev;
      head.prev = null;
      return head.value;
    }
  }

  T? peek() {
    return _head?.value;
  }
}

void main(List<String> args) {
  Stack<int> stack = Stack();
  print(stack.peek());
  print(stack.length);
  stack.pop();
  print(stack.peek());
  print(stack.length);
  stack.push(2);
  print(stack.peek());
  print(stack.length);
  stack.push(4);
  stack.push(15);
  print(stack.length);
  print(stack.pop());
  print(stack.pop());
  print(stack.pop());
  print(stack.peek());
  print(stack.pop());
  print(stack.length);
}
